clients/telemetry-kafka/src/main/java/org/eclipse/hono/client/telemetry/kafka/AbstractKafkaBasedDownstreamSender.java

```diff
@@ -154,14 +156,34 @@ public AbstractKafkaBasedDownstreamSender(
             }
         }
 
-        // change TTL from seconds to milliseconds (according to API specification)
-        headerProperties.computeIfPresent(MessageHelper.SYS_HEADER_PROPERTY_TTL, (k, v) -> {
-            if (v instanceof Number) {
-                return ((Number) v).longValue() * 1000L;
+        // make sure that device provided TTL is capped at max TTL (if set)
+        headerProperties.compute(MessageHelper.SYS_HEADER_PROPERTY_TTL, (k, v) -> {
+
+            final long maxTtl = Optional.ofNullable(tenant)
+                    .flatMap(t -> Optional.ofNullable(t.getResourceLimits()))
+                    .map(ResourceLimits::getMaxTtl)
+                    .orElse(TenantConstants.UNLIMITED_TTL);
+
+            final long ttlSeconds;
+            if (v instanceof Number) { // TTL is configured
+                final long ttl = ((Number) v).longValue();
+                if (maxTtl != TenantConstants.UNLIMITED_TTL && ttl > maxTtl) {
+                    log.debug("limiting TTL [{}s] to max TTL [{}s]", ttl, maxTtl);
+                    ttlSeconds = maxTtl;
+                } else {
+                    log.trace("keeping message TTL [{}s, max TTL: {}s]", ttl, maxTtl);
+                    ttlSeconds = ttl;
+                }
             } else {
-                log.error("TTL property is not a number [{}]", v);
-                return v;
+                if (maxTtl != TenantConstants.UNLIMITED_TTL) {
+                    log.debug("setting TTL to tenant's max TTL [{}s]", maxTtl);
+                    ttlSeconds = maxTtl;
+                } else {
+                    log.error("TTL property is not a number [{}]", v);
```

Comment:
this is also true in the case above, isn't it?

Reply:
Yes. It was a mistake to add this log statement. The value `v` is especially not a number if it is `null`.